Switch from `jsonpath_lib` to `jsonpath-rust`

ilya-bobyr opened the pull request:
## Motivation
`jsonpath_lib` is unmaintained, which is not ideal.  It is also forcing everyone to enable a `preserve_order` feature in the `serde_json` dependency.

`preserve_order` is not following the proper feature model, where features are only adding functionality.  It is actually changing how `serde_json` works, switching from `BTreeMap` to `IndexMap`.

## Solution
Switch from [`jsonpath_lib`](https://crates.io/crates/jsonpath_lib) to [`jsonpath-rust`](https://crates.io/crates/jsonpath-rust).

Change has been tested with both unit and integration tests.

---

Based on https://github.com/gregcusack/kube/pull/1

Dav1dde commented on examples/dynamic_jsonpath.rs:
```diff
@@ -13,19 +16,26 @@ async fn main() -> anyhow::Result<()> {
-    );
+    let jsonpath = {
+        let path = std::env::var("JSONPATH").unwrap_or_else(|_| ".items[*].spec.containers[*].image".into());
+        JsonPathInst::from_str(&format!("{}{}", "$", path.clone()))
```
```suggestion
        format!("${path}").parse::<JsonPathInst>()
```

Also means you can get rid of the `FromStr` import

ilya-bobyr replied:
Applied.

Dav1dde commented on kube-client/src/client/auth/mod.rs:
```diff
@@ -474,22 +475,32 @@ fn token_from_gcp_provider(provider: &AuthProviderConfig) -> Result<ProviderToke
+fn extract_value(json: &serde_json::Value, context: &str, path: &str) -> Result<String, Error> {
+    let parsed_path = {
+        let trimmed = path.trim_matches(|c| c == '"' || c == '{' || c == '}');
+        JsonPathInst::from_str(trimmed).map_err(|err| {
```
Same here `.parse()` then you can also nicely chain it with the trim above

ilya-bobyr replied:
Done.

Dav1dde reviewed (commented):
I did not test it / look for functional changes

Dav1dde commented on examples/dynamic_jsonpath.rs:
```diff
@@ -13,19 +15,27 @@ async fn main() -> anyhow::Result<()> {
-    );
+    let jsonpath = {
+        let path = std::env::var("JSONPATH").unwrap_or_else(|_| ".items[*].spec.containers[*].image".into());
+        format!("{}{}", "$", path.clone())
```
Can still do this:

```suggestion
        format!("${path}")
```

ilya-bobyr replied:
Pre-existing code :P
Fixed :)

Dav1dde commented on kube-client/src/client/auth/mod.rs:
```diff
@@ -474,22 +474,32 @@ fn token_from_gcp_provider(provider: &AuthProviderConfig) -> Result<ProviderToke
+        Ok(val.to_owned())
+    } else {
+        Err(Error::AuthExec(format!(
+            "Target {:?} value {:?} is not a string: {:?}",
```
```suggestion
            "Target {context:?} value {path:?} is not a string: {res:?}",
```

ilya-bobyr replied:
Yes, `res` needs to be dereferenced.
I think, it is confusing when inline symbols are mixed with expressions that are passed as arguments.
Even clippy only suggests inlining only if everything can be inlined.
Controlled by [`allow-mixed-uninlined-format-args`](https://doc.rust-lang.org/nightly/clippy/lint_configuration.html#allow-mixed-uninlined-format-args).

But, I can make this change if this is the style this project uses.

Dav1dde replied:
Fair point. Personally I think it helps with readability, especially since it is only the last argument.

I looked for a precedent in the code base and couldn't find one (only a bunch of old stuff which now could be inlined).

So I am fine with it either way.

Dav1dde reviewed (commented):
Just some small `format!` nits other than that LGTM 👍 

Out of curiosity have you looked into https://docs.rs/serde_json_path/latest/serde_json_path/ - I have used it before and quite liked it.

clux commented on kube-client/src/client/auth/mod.rs:
```diff
@@ -474,22 +474,32 @@ fn token_from_gcp_provider(provider: &AuthProviderConfig) -> Result<ProviderToke
+            ))
+        })?;
+
+    let res = parsed_path.find_slice(json);
+
+    let Some(res) = res.into_iter().next() else {
+        return Err(Error::AuthExec(format!(
+            "Target {context:?} value {path:?} not found"
+        )));
+    };
```
will it always work to use `find_slice` here? to me that sounds like we expect the query to return a vector of things, but curious if this will work if we do a query returning a single item like `'{.items[3].metadata.name}'`

gregcusack replied:
ya so `find_slice` will return an empty vector if the query doesn't match and will return anything else as elements in an array, including if it results in just a single item. The item will just be returned in the vector. See: https://github.com/besok/jsonpath-rust/blob/main/src/lib.rs#L450-L460. 

Test here shows find_slice() returning a single item in a vector: https://github.com/besok/jsonpath-rust/blob/main/src/lib.rs#L1194-L1198

ilya-bobyr replied:
Existing logic here is a bit suspicious.
If any elements are selected by the JSONPath, it would always use the first element, even if there is more than one result.

I wonder if it would be better to fail with an error, if there are multiple results.
If someone really wants to use the first result, they can always clarify their selector expression.
On the other hand, choosing the first result could hide an error and complicate debugging.

I can send a separate PR to change this behavior.

clux replied:
I guess this is ultimately only for a small amount of niche behaviour; local auth where the usage from exec based providers needs to grab one key - so it's not as exposed as something like kubectl jsonpath querying. If there's a clear way to improve it feel free to raise a PR, but I don't see this as a big problem atm.

ilya-bobyr commented:
Replaced

```rust
    format!("{}{}", $, path.clone())
```
with
```rust
    format!("${path}")
```

and rebased on top of the latest `main`.

ilya-bobyr commented:
> Out of curiosity have you looked into [docs.rs/serde_json_path/latest/serde_json_path](https://docs.rs/serde_json_path/latest/serde_json_path/) - I have used it before and quite liked it.

[`jsonpath-rust`](https://crates.io/crates/jsonpath-rust) seems to be much more popular, compared to [`serde_json_path`](https://crates.io/crates/serde_json_path), judging just by the download count.
And, similarly, GitHub stars and fork counts for [`jsonpath-rust`](https://github.com/besok/jsonpath-rust) are higher, compared to [`serde_json_path`](https://github.com/hiltontj/serde_json_path).

It seems that both libraries can work directly with [`serde_json::Value`](https://docs.rs/serde_json/1.0.108/x86_64-unknown-linux-gnu/serde_json/value/enum.Value.html) objects.  But `jsonpath-rust` was developed several years earlier.  Not sure what are the advantages of `serde_json_path`.

I have not used any Rust JSONPath libraries much.  And the scope of this PR is that we have a problem with the outdated dependency, that is braking our dependency tree.  It is not that I prefer one library over another :)

clux commented:
Given how little usage there is of the `serde_json_path` variant, it is probably too fresh for us to bundle anyway. This takes care of the out-of-date problem for now. Happy to take up the discussion of one-vs-the-other again in the future once the ecosystem stabilises (they have both been releasing new versions quite recently).

Btw, this needs an MSRV bump to pass CI which is perfectly reasonable for us to do at this stage. Have done this in #1353

Dav1dde commented:
I think this is the right choice, was just (personally) curious if you explored more alternatives 👍 .
